File: tools/parity/dispatch/browser-search.ts
// TS dispatch for the browser-search parity module: maps the shared vector
// function names to the real `src/shared/browser-url.ts` exports so the harness
// compares the live TS reference against the Rust port. Only the self-contained
// pure functions are covered; the Kagi session-link / navigation normaliser is
// deferred in the Rust port.

import { buildSearchUrl, looksLikeSearchQuery, type SearchEngine } from '../../../src/shared/browser-url'

export function dispatch(fn: string, input: unknown): unknown {
  switch (fn) {
    case 'buildSearchUrl': {
      const { query, engine } = input as { query: string; engine?: SearchEngine }
      // Pass engine through; undefined triggers the `engine = DEFAULT_SEARCH_ENGINE` default param.
      return buildSearchUrl(query, engine)
    }
    case 'looksLikeSearchQuery':
      return looksLikeSearchQuery(input as string)
    default:
      throw new Error(`unknown function ${fn}`)
  }
}

File: rust/crates/orca-dispatch/src/modules/browser_search.rs
//! Parity dispatch for `orca_core::browser_search` vs the self-contained pure
//! search-query parts of `src/shared/browser-url.ts`. The Kagi session-link /
//! navigation normaliser is deferred in the Rust port, so only `buildSearchUrl`
//! (without options) and `looksLikeSearchQuery` are covered here.

use orca_core::browser_search::{
    build_search_url, looks_like_search_query, SearchEngine, DEFAULT_SEARCH_ENGINE,
};
use serde_json::{json, Value};

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        "buildSearchUrl" => {
            let query = input.get("query").and_then(Value::as_str).unwrap_or("");
            let engine = parse_engine(input.get("engine"));
            Value::String(build_search_url(query, engine))
        }
        "looksLikeSearchQuery" => Value::Bool(looks_like_search_query(input.as_str().unwrap_or(""))),
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}

/// Map the TS `SearchEngine` string ids to the Rust enum. Absent/undefined falls
/// back to the default engine, mirroring TS's `engine = DEFAULT_SEARCH_ENGINE`.
fn parse_engine(value: Option<&Value>) -> SearchEngine {
    match value.and_then(Value::as_str) {
        Some("google") => SearchEngine::Google,
        Some("duckduckgo") => SearchEngine::DuckDuckGo,
        Some("bing") => SearchEngine::Bing,
        Some("kagi") => SearchEngine::Kagi,
        _ => DEFAULT_SEARCH_ENGINE,
    }
}
